src/hotspot/share/logging/logAsyncFlusher.cpp

```diff
@@ -0,0 +1,233 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+#include "precompiled.hpp"
+#include "jvm.h"
+#include "logging/logAsyncFlusher.hpp"
+#include "logging/logConfiguration.hpp"
+#include "logging/logFileOutput.hpp"
+#include "logging/logHandle.hpp"
+#include "runtime/atomic.hpp"
+
+void AsyncLogMessage::writeback() {
+  if (_message != NULL) {
+    _output.write_blocking(_decorations, _message);
+    os::free(_message);
+  }
+}
+
+void LogAsyncFlusher::enqueue_impl(const AsyncLogMessage& msg) {
+  assert_lock_strong(&_lock);
+
+  if (_buffer.size() >= _buffer_max_size)  {
+    const AsyncLogMessage* h = _buffer.front();
+    assert(h != NULL, "sanity check");
+
+    if (h->message() != nullptr) {
+      bool p_created;
+      uint32_t* counter = _stats.add_if_absent(h->output(), 0, &p_created);
+      *counter = *counter + 1;
+
+      if (Verbose) {
+        // Temporarily turn off SerializeVMOutput so defaultStream will not
+        // invoke set_owner(self) for tty_lock.
+        FlagSetting t(SerializeVMOutput, false);
+        // The writing below can not guarantee non-blocking because tty may be piped by the filesystems
+        // or throttled by XOFF, so only dump the dropping message in Verbose mode.
+        tty->print_cr("asynclog dropping message: %s", h->message());
+      }
+    }
+
+    _buffer.pop_front();
+  }
+  assert(_buffer.size() < _buffer_max_size, "_buffer is over-sized.");
+  _buffer.push_back(msg);
+
+  // notify asynclog thread if occupancy is over 3/4
+  size_t sz = _buffer.size();
+  if (sz > (_buffer_max_size >> 2) * 3 ) {
+    _lock.notify();
+  }
+}
+
+void LogAsyncFlusher::enqueue(LogFileOutput& output, const LogDecorations& decorations, const char* msg) {
+  AsyncLogMessage m(output, decorations, os::strdup(msg));
+
+  { // critical area
+    // The rank of _lock is same as _tty_lock on purpuse.
+    // if logging thread is holding _tty_lock now, temporarily yield to _lock.
+    ttyUnlocker ttyul;
+    MutexLocker ml(&_lock, Mutex::_no_safepoint_check_flag);
+    enqueue_impl(m);
+  }
+}
+
+// LogMessageBuffer consists of a multiple-part/multiple-line messsages.
+// the mutex here gurantees its integrity.
+void LogAsyncFlusher::enqueue(LogFileOutput& output, LogMessageBuffer::Iterator msg_iterator) {
+  ttyUnlocker ttyul;
+  MutexLocker ml(&_lock, Mutex::_no_safepoint_check_flag);
+
+  for (; !msg_iterator.is_at_end(); msg_iterator++) {
+    AsyncLogMessage m(output, msg_iterator.decorations(), os::strdup(msg_iterator.message()));
+    enqueue_impl(m);
+  }
+}
+
+LogAsyncFlusher::LogAsyncFlusher()
+  : _state(ThreadState::Running),
+    _lock(Mutex::tty, "async-log-monitor", true /* allow_vm_block */, Mutex::_safepoint_check_never),
+    _stats(17 /*table_size*/) {
+  if (os::create_thread(this, os::asynclog_thread)) {
+    os::start_thread(this);
```

Comment:
Here is some error handling needed if thread creation failed. UL should drop to sync mode then.

Reply:
make sense! I will take a look at the error handling.